js/copy-service/Evaluator/Evaluator.ts

```diff
@@ -0,0 +1,109 @@
+import _ from 'lodash';
+
+import CopyService, { AST } from '../CopyService';
+import ErrorHandler from '../ErrorHandler/ErrorHandler';
+import IntlCopyService from '../IntlCopyService';
+import Substitutions from '../Substitutions/Substitutions';
+
+/**
+ * Provides an interface to recursively generate copy evaluated with substitutions.
+ *
+ * @interface
+ */
+class Evaluator {
+  /**
+   * The copy service used for evaluation.
+   *
+   * @type {CopyService|IntlCopyService}
+   */
+  copyService: CopyService | IntlCopyService;
+
+  /**
+   * A cache for previously evaluated cope.
+   *
+   * @type {WeakMap}
+   */
+  evaluationCache: WeakMap<object, unknown>;
+
+  /**
+   * When `true`, the functional ^{}{} syntax will run the configured method when evaluating.
+   * When `false`, the copy will be returned without passing it through the function.
+   *
+   * @type {boolean}
+   */
+  allowFunctional: boolean;
+
+  constructor(
+    copyService: CopyService | IntlCopyService,
+    params: { allowFunctional: boolean; } = { allowFunctional: true }
+  ) {
+    this.copyService = copyService;
+
+    this.evaluationCache = new WeakMap();
+
+    this.allowFunctional = params.allowFunctional;
+  }
+
+  /**
+   * Gets the cached evaluation result for the given ast, if it exists.
+   */
+  getCached(ast: AST): unknown | null {
```

Comment:
See other comment on the cache/WeakMap type.

```suggestion
  getCached(ast: SyntaxNode): any {
```